docker/importer/importer.py

```diff
@@ -455,6 +481,104 @@ def thread_init():
+    """Process deletions from a GCS bucket source.
+
+    This validates the continued existence of every Bug in Datastore (for the
+    given source) against every bug currently in that source's GCS bucket,
+    calculating the delta. The bugs determined to have been
+    deleted from GCS are then flagged for treatment by the worker.
+
+    If the delta is too large, something undesirable has been assumed to have
+    happened and treatment is aborted.
+    """
+
+    logging.info("Begin processing bucket for deletions: %s", source_repo.name)
+
+    # Get all the existing non-withdrawn Bug IDs for
+    # source_repo.name in Datastore
+    query = osv.Bug.query()
+    query = query.filter(osv.Bug.source == source_repo.name)
+    result = list(query.fetch(keys_only=False))
+    result.sort(key=lambda r: r.id())
+    VulnAndSource = namedtuple('VulnAndSource', ['id', 'path'])
+    vuln_ids_for_source = [
+        VulnAndSource(id=r.id(), path=r.source_id.split(':')[1])
+        for r in result
+        if not r.withdrawn
+    ]
+
+    storage_client = storage.Client()
+    # Get all of the existing records in the GCS bucket
+    # (to get their IDs for checking against Datastore)
+    listed_blobs = list(
+        storage_client.list_blobs(
+            source_repo.bucket, prefix=source_repo.directory_path))
+
+    import_failure_logs = []
+
+    def convert_blob_to_vulns(blob: storage.Blob) -> Optional[Tuple[str]]:
+      """Download and parse GCS blob into [vuln.id]"""
+      if not _is_vulnerability_file(source_repo, blob.name):
+        return None
+
+      # Use the _client_store thread local variable
+      # set in the thread pool initializer
+      # Download in a blob generation agnostic way to cope with the file
+      # changing between when it was listed and now.
+      blob_bytes = storage.Blob(
+          blob.name, blob.bucket,
+          generation=None).download_as_bytes(_client_store.storage_client)
+
+      vuln_ids = []
+      try:
+        vulns = osv.parse_vulnerabilities_from_data(
+            blob_bytes,
+            os.path.splitext(blob.name)[1],
+            strict=self._strict_validation)
+        for vuln in vulns:
+          vuln_ids.append(vuln.id)
+
+        return vuln_ids
+      except Exception as e:
+        logging.error('Failed to parse vulnerability %s: %s', blob.name, e)
+        # Don't include error stack trace as that might leak sensitive info
+        # List.append() is atomic and threadsafe.
+        import_failure_logs.append('Failed to parse vulnerability "' +
+                                   blob.name + '"')
+        return None
+
+    # Setup storage client
+    def thread_init():
+      _client_store.storage_client = storage.Client()
+
+    with ThreadPoolExecutor(
+        _BUCKET_THREAD_COUNT, initializer=thread_init) as executor:
+      vulns_in_gcs = [
+          v for v in executor.map(convert_blob_to_vulns, listed_blobs) if v
```

Comment:
Is this correct? 

`convert_blobs_to_vulns` returns a list[str] now, which means that `vulns_in_gcs` is now a list of lists? 

Do we need to concatenate all the results here instead? i.e. `sum(v for v in executor.map(convert_blob_to_vulns, listed_blobs), [])`

Reply:
This has been overhauled in fabd741, so the comment may not still be applicable. I'm still testing the current code.

Reply:
I think this is no longer relevant.